consumer/pdatautil/metricsbuilder.go

```diff
@@ -0,0 +1,205 @@
+// to be much easier and more fluent than than using pdata.Metrics directly.
+// All of the methods on it except for Flush() will return the same instance of
+// MetricsBuilder as a return value, allowing you to chain method calls easily,
+// similar to the Java builder pattern.
+//
+// The MetricsBuilder is designed for cases where receivers are generating
+// metrics from scratch, where generally you will have a single datapoint per
+// metric/label combination.
+//
+// It is NOT thread-safe, so you should use an external mutex if using it from
+// multiple goroutines.
+type MetricsBuilder struct {
+	// If set, this instrumentation library name will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryName string
+	// If set, this instrumentation library version will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryVersion string
+	// These attributes will be added to the Resource object on all
+	// ResourceMetrics instances created by the builder.
+	ResourceAttributes map[string]string
+	// This time will be used as the Timestamp for all metrics generated.  It
+	// can be updated with a new timestamp at any time.
+	Timestamp time.Time
+
+	metrics pdata.MetricSlice
+	labels  map[string]string
+	isInit  bool
+}
+
+// Flush creates a pdata.Metrics instance with all of the pending metrics on
+// the builder and returns it.  Subsequent calls to Flush will only return
+// metrics added after the previous Flush call.  Labels set with With*Labels
+// will be retained after a call to Flush.
+func (mb *MetricsBuilder) Flush() pdata.Metrics {
+	md := data.NewMetricData()
+	rms := md.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+	rm.InitEmpty()
+
+	res := rm.Resource()
+	res.InitEmpty()
+	for k, v := range mb.ResourceAttributes {
+		res.Attributes().Insert(k, pdata.NewAttributeValueString(v))
+	}
+
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	ilm := ilms.At(0)
+	ilm.InitEmpty()
+
+	il := ilm.InstrumentationLibrary()
+	il.InitEmpty()
+	il.SetName(mb.InstrumentationLibraryName)
+	il.SetVersion(mb.InstrumentationLibraryVersion)
+
+	mb.metrics.MoveAndAppendTo(ilm.Metrics())
+
+	mb.init()
+
+	return pdata.Metrics{InternalOpaque: md}
+}
+
+func (mb *MetricsBuilder) init() {
+	mb.metrics = pdata.NewMetricSlice()
+	mb.isInit = true
+}
+
+// WithLabels sets a set of labels on the builder. All subsequent calls to
+// create metrics will create metrics that use the most recently set of labels
+// set with this method.  The input map's entries are copied so the map can be
+// mutated freely without affecting the builder.
+func (mb *MetricsBuilder) WithLabels(l map[string]string) *MetricsBuilder {
+	mb.labels = nil
+	return mb.WithAdditionalLabels(l)
+}
+
+// WithAdditionalLabels will add the given labels to the already existing label
+// set, instead of replacing them outright.  The input map's entries are copied
```

Comment:
Do we really need the copying semantics? It is a performance hit for the WithLabels() case. Can we require the caller give up the ownership of the map and just copy the reference to the map in WithLabels().
(I don't know if performance matters for this call, I assume it may be called a lot when generating metrics, but I may be wrong and it is insignificant in which case copying is indeed safer and preferable).

Reply:
Given how many bugs I have run into in the Smart Agent that stemmed from mutating a map that was passed around to different goroutines, I have found it better to just copy maps and take a small performance hit and totally avoid the risk of map induced panics.  Since Go has no notion of ownership in the compiler (or test utilities that I'm aware of), so there is no way to programatically ensure something truly "gives up ownership" and stays that way.